### PR TITLE
stm32mp1: prepare for SiP SMC services

### PR DESCRIPTION
**Pull request by etienne-lms**
Implement secure monitor platform handlers foundations for
platform stm32mp1 to handle SiP SMC services.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

### CODE REVIEW

File: core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	struct smc_args *args = (void *)&ctx->nsec.r0;
+
+	COMPILE_TIME_ASSERT(SM_NSEC_CTX_OFFSET(r0) == SMC_ARGS_OFFSET(a0) &&
+			    SM_NSEC_CTX_OFFSET(r1) == SMC_ARGS_OFFSET(a1) &&
+			    SM_NSEC_CTX_OFFSET(r2) == SMC_ARGS_OFFSET(a2) &&
+			    SM_NSEC_CTX_OFFSET(r3) == SMC_ARGS_OFFSET(a3) &&
+			    SM_NSEC_CTX_OFFSET(r4) == SMC_ARGS_OFFSET(a4) &&
+			    SM_NSEC_CTX_OFFSET(r5) == SMC_ARGS_OFFSET(a5) &&
+			    SM_NSEC_CTX_OFFSET(r6) == SMC_ARGS_OFFSET(a6) &&
+			    SM_NSEC_CTX_OFFSET(r7) == SMC_ARGS_OFFSET(a7));
```
Comment (etienne-lms) [inline core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c:53]: Related to comment https://github.com/OP-TEE/optee_os/pull/3736/commits/58b3a57dfcbcb397387de872b9e7f1db6a2a9ac6#r406267428 in the P-R: this assertion was to ensure `struct sm_nsec_ctx`:`r0`..`r7`  do match fields struct thread_smc_args`:`a0`..`a7`. 
Is it really necessary?

There is a cast in [**sm.c**](https://github.com/OP-TEE/optee_os/blob/3.8.0/core/arch/arm/sm/sm.c#L18-L36) that implicitly expect this mapping:
```c
uint32_t sm_from_nsec(struct sm_ctx *ctx)
{
	uint32_t *nsec_r0 = (uint32_t *)(&ctx->nsec.r0);

	(...)

		smc_std_handler((struct thread_smc_args *)nsec_r0, &ctx->nsec);
}
```
Do you suggest to do the same here, removing the assertion.

File: core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ * SIP function STM32_SIP_SVC_FUNC_SCMI_AGENT0/1
+ *
+ * Process SCMI message pending in SCMI shared memory buffer.
+ * No input or output arguments through CPU general purpose registers.
+ */
+#define STM32_SIP_SVC_FUNC_SCMI_AGENT0		0x2000
+#define STM32_SIP_SVC_FUNC_SCMI_AGENT1		0x2001
+
+/* Generic input/output arguments for an SMC */
+struct smc_args {
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h:76]: yes sure, obviously. i will fix as suggested.

File: core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#define STM32_SIP_SVC_FUNC_VERSION		0xff03
+
+/*
+ * SIP function STM32_SIP_SVC_FUNC_SCMI_AGENT0/1
+ *
+ * Process SCMI message pending in SCMI shared memory buffer.
+ * No input or output arguments through CPU general purpose registers.
+ */
+#define STM32_SIP_SVC_FUNC_SCMI_AGENT0		0x2000
+#define STM32_SIP_SVC_FUNC_SCMI_AGENT1		0x2001
```
Comment (etienne-lms) [inline core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h:73]: Oh, these were planned to be defined of a later change, but maybe I can keep them in this P-R.

File: core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c:54]: Yes, I think that's safe (to remove the assertion).

File: core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h (hunk 3 above)
Comment (jenswi-linaro) [reply core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h:75]: Sure

**Comment by etienne-lms** [comment]
updated

**Comment by etienne-lms** [comment]
squashed and tag applied.

**Pull request merged**